src/doc/unstable-book/src/language-features/unsized-locals.md

```diff
@@ -0,0 +1,199 @@
+
+[RFC1909]: https://github.com/rust-lang/rfcs/blob/master/text/1909-coercions.md
+
+```rust
+#![feature(unsized_locals)]
+
+use std::any::Any;
+
+fn main() {
+    let x: Box<Any> = Box::new(42);
+    let x: Any = *x;
+    //  ^ unsized local variable
+    //           ^^ unsized temporary
+    foo(x);
+}
+
+fn foo(_: Any) {}
+//     ^^^^^^ unsized argument
+```
+
+The RFC still forbids the following unsized expressions:
+
+```rust,ignore
+#![feature(unsized_locals)]
+
+use std::any::Any;
+
+struct MyStruct<T: ?Sized> {
+    content: T,
+}
+
+struct MyTupleStruct<T: ?Sized>(T);
+
+fn answer() -> Box<Any> {
+    Box::new(42)
+}
+
+fn main() {
+    // You CANNOT have unsized statics.
+    static X: Any = *answer();  // ERROR
+    const Y: Any = *answer();  // ERROR
+
+    // You CANNOT have struct initialized unsized.
+    MyStruct { content: *answer() };  // ERROR
+    MyTupleStruct(*answer());  // ERROR
+    (42, *answer());  // ERROR
+
+    // You CANNOT have unsized return types.
+    fn my_function() -> Any { *answer() }  // ERROR
+
+    // You CAN have unsized local variables...
+    let mut x: Any = *answer();  // OK
+    // ...but you CANNOT reassign to them.
+    x = *answer();  // ERROR
+
+    // You CANNOT even initialize them separately.
+    let y: Any;  // OK
+    y = *answer();  // ERROR
+
+    // Not mentioned in the RFC, but by-move captured variables are also Sized.
+    let x: Any = *answer();
+    (move || {  // ERROR
+        let y = x;
+    })();
+
+    // You CAN create a closure with unsized arguments,
+    // but you CANNOT call it.
+    // This is an implementation detail and may be changed in the future.
+    let f = |x: Any| {};
+    f(*answer());  // ERROR
+}
+```
+
+However, the current implementation allows `MyTupleStruct(..)` to be unsized. This will be fixed in the future.
+
+## By-value trait objects
+
+With this feature, you can have by-value `self` arguments without `Self: Sized` bounds.
+
+```rust
+#![feature(unsized_locals)]
+
+trait Foo {
+    fn foo(self) {}
+}
+
+impl<T: ?Sized> Foo for T {}
+
+fn main() {
+    let slice: Box<[i32]> = Box::new([1, 2, 3]);
+    <[i32] as Foo>::foo(*slice);
+}
+```
+
+And `Foo` will also be object-safe. However, this object-safety is not yet implemented.
+
+```rust,ignore
+#![feature(unsized_locals)]
+
+trait Foo {
+    fn foo(self) {}
+}
+
+impl<T: ?Sized> Foo for T {}
+
+fn main () {
+    let slice: Box<Foo> = Box::new([1, 2, 3]);
+    // doesn't compile yet
+    <Foo as Foo>::foo(*slice);
+}
+```
+
+Unfortunately, this is not implemented yet.
+
+One of the objectives of this feature is to allow `Box<FnOnce>`, instead of `Box<FnBox>` in the future. See [#28796] for details.
+
+[#28796]: https://github.com/rust-lang/rust/issues/28796
+
+## Variable length arrays
+
+The RFC also describes an extension to the array literal syntax `[e; n]`: you'll be able to specify non-const `n` to allocate variable length arrays on the stack.
```

Comment:
I remember the syntax `[e; dyn n]` was the preferred one.

Reply:
The [RFC](https://github.com/rust-lang/rfcs/blob/master/text/1909-unsized-rvalues.md#the-bikeshed) prefers `[e; n]` (where the array is considered dyn if `n` captures a variable).

Reply:
Ok. It seems that many people prefer `dyn` syntax in the pull request. But it is not accepted as final decision.

Reply:
+1 -- my impression was that everyone in the RFC thread was in support of *allowing* `dyn` here, with some thinking that it should *only* work with `dyn`. IMO we should support `dyn` and lint against not using it (this is the current policy for `dyn` in trait objects).

Reply:
The proposal to merge the RFC was left subject to [remaining questions about the syntax](https://github.com/rust-lang/rfcs/pull/1909#issuecomment-327872398), after which @nikomatsakis [responded in favor of `dyn`](https://github.com/rust-lang/rfcs/pull/1909#issuecomment-328976827).

Reply:
Then it seems good to edit the unstable book to say something like "`[e; n]` or `[e; dyn n]`".

Reply:
+1 for the `dyn` syntax, at least for now. Less moving parts.